ui/raidboss/data/06-ew/trial/zodiark-ex.ts

```diff
@@ -1,23 +1,297 @@
+      type: 'AddedCombatant',
+      netRegex: NetRegexes.addedCombatant({ name: 'Roiling Darkness', capture: false }),
+      suppressSeconds: 1,
+      alertText: (_data, _matches, output) => output.text!(),
+      outputStrings: {
+        text: {
+          en: Outputs.killAdds.en + '(back first)',
+        },
+      },
+    },
+    {
+      // 67E4 Green Beam, 67E5 Rectangle, 67E6 Wedge
+      id: 'ZodiarkEx Arcane Sigil Start',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: ['67E4', '67E5', '67E6'], source: 'Arcane Sigil' }),
+      run: (data, matches, _output) => {
+        if (parseFloat(matches.y) < 100)
+          data.activeFrontSigils.push({ x: parseFloat(matches.x), y: parseFloat(matches.y), typeId: matches.id, npcId: matches.sourceId });
+      },
+    },
+    {
+      id: 'ZodiarkEx Arcane Sigil Start Cleanup',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: ['67E4', '67E5', '67E6'], source: 'Arcane Sigil', capture: false }),
+      delaySeconds: 0.2,
+      suppressSeconds: 0.5,
+      alertText: (data, _matches, output) => {
+        const activeFrontSigils = data.activeFrontSigils;
+        data.activeFrontSigils = [];
+        if (activeFrontSigils.length === 1 && activeFrontSigils[0]?.typeId === '67E4')
+          return output.sides!();
+        if (activeFrontSigils.length === 1 && activeFrontSigils[0]?.typeId === '67E5')
+          return output.south!();
+        if (activeFrontSigils.length === 1 && activeFrontSigils[0]?.typeId === '67E6')
+          return output.north!();
+        if (activeFrontSigils.length === 2 && activeFrontSigils[0]?.typeId === '67E4' && activeFrontSigils[1]?.typeId === '67E4')
+          return output.middle!();
+        if (activeFrontSigils.length === 3) {
+          for (const sig of activeFrontSigils) {
+            // Find the middle sigil
+            if (sig.x > 90 && sig.x < 110) {
+              if (sig.typeId === '67E4')
+                return output.frontsides!();
+              if (sig.typeId === '67E5')
+                return output.backmiddle!();
+              if (sig.typeId === '67E6')
+                return output.frontmiddle!();
+            }
+          }
+        }
+      },
+      outputStrings: {
+        south: Outputs.south,
+        north: Outputs.north,
+        frontsides: {
+          en: 'front sides',
+        },
+        backmiddle: {
+          en: 'back middle',
+        },
+        frontmiddle: {
+          en: 'front middle',
+        },
+        sides: Outputs.sides,
+        middle: Outputs.middle,
+      },
+    },
+    {
+      // 67EC is leaning left, 67ED is leaning right
+      id: 'ZodiarkEx Algedon',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: ['67EC', '67ED'], source: 'Zodiark' }),
+      alertText: (data, matches, output) => {
+        if (matches.id === '67EC') {
+          // NE/SW
+          return output.combo!({ first: 'Northeast', second: 'Southwest' });
```

Comment:
```suggestion
          return output.combo!({ first: output.northeast!(), second: output.southwest!() });
```

and same for the one below

Reply:
I haven't really understood how that !() magic works, I guess it somehow gets the correct string out of the en/de/fr/... object?

Regardless I've fixed it and will research that later. :3

Reply:
`output` is a [JavaScript proxy object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy).  Every field in `outputStrings` is a property of `output` that returns a function (that can take optional parameters).  TypeScript doesn't know about proxies, so the `!` is there to just say "trust me, I know this is not null".

https://github.com/quisquous/cactbot/blob/main/ui/raidboss/popup-text.ts#L233-L368 is the implementation of `Output` if that's helpful.  But, it's the indirection between `outputStrings` and `output` that allows users to override the output strings.